src/core/worker_loop_.ml

```diff
@@ -0,0 +1,153 @@
+open Types_
+
+type fiber = Picos.Fiber.t
+
+type task_full =
+  | T_start of {
+      fiber: fiber;
+      f: unit -> unit;
+    }
+  | T_resume : {
+      fiber: fiber;
+      k: unit -> unit;
+    }
+      -> task_full
+
+type around_task =
+  | AT_pair : (Runner.t -> 'a) * (Runner.t -> 'a -> unit) -> around_task
+
+exception No_more_tasks
+
+type 'st ops = {
+  schedule: 'st -> task_full -> unit;
+  get_next_task: 'st -> task_full;  (** @raise No_more_tasks *)
+  get_thread_state: unit -> 'st;
+      (** Access current thread's worker state from any worker *)
+  around_task: 'st -> around_task;
+  on_exn: 'st -> Exn_bt.t -> unit;
+  runner: 'st -> Runner.t;
+  before_start: 'st -> unit;
+  cleanup: 'st -> unit;
+}
+
+(** A dummy task. *)
+let _dummy_task : task_full = T_start { f = ignore; fiber = _dummy_fiber }
+
+[@@@ifge 5.0]
+
+let[@inline] discontinue k exn =
+  let bt = Printexc.get_raw_backtrace () in
+  Effect.Deep.discontinue_with_backtrace k exn bt
+
+let with_handler (type st arg) ~(ops : st ops) (self : st) :
+    (unit -> unit) -> unit =
+  let current =
+    Some
+      (fun k ->
+        match get_current_fiber_exn () with
+        | fiber -> Effect.Deep.continue k fiber
+        | exception exn -> discontinue k exn)
+  and yield =
+    Some
+      (fun k ->
+        let fiber = get_current_fiber_exn () in
+        match
+          let k () = Effect.Deep.continue k () in
+          ops.schedule self @@ T_resume { fiber; k }
+        with
+        | () -> ()
+        | exception exn -> discontinue k exn)
+  and reschedule trigger fiber k : unit =
+    ignore (Picos.Fiber.unsuspend fiber trigger : bool);
+    let k () = Picos.Fiber.resume fiber k in
+    let task = T_resume { fiber; k } in
+    ops.schedule self task
```

Comment:
I don't know enough about Picos fibers, but this looks like there could be a race here, when `fiber` finishes before `ops.schedule` runs.